design-proposals/golden-image-delivery-and-update-pipeline.md

```diff
@@ -0,0 +1,287 @@
+
+A **DataSource** object references an import/clone source for a PVC. Functionally, it is similar to the 'source' field of a DataVolume. The DataSource’s spec is mutable, which has the side effect of allowing the DataSource to act as a link that will redirect to whatever source the link currently points to. In the future, the DataSource will act as a PVC Populator as well as a DataVolume source.
+
+**DataImportCron can manage a corresponding DataSource** to always point to the latest most up-to-date PVC that the cron has created. The DataImportCron should also be capable of adopting and taking ownership of an existing DataSource in order to statisfy the condition where an admin wishs to opt-in to using auto delivery/updates later on.
+
+
+## CDI: DataVolume DataSource Source Field
+
+The DataVolume object will add a new field capable of importing pvc content using a DataSource. 
+
+![DataVolume DataSource Field](golden-image-delivery-images/volume-reference-source.png "DataVolume DataSource Field")
+
+## SSP: Setup DataImportCrons for common templates
+
+Since the SSP operator manages installing common templates, this operator should also be responsible for setting up the DataImportCrons used to automatically source the disk images into the cluster that the common templates use.
+
+The location the DataImportCrons pull from will be configurable on the SSP operators Custom Resource.
+
+![SSP Disk Import Pipelines](golden-image-delivery-images/ssp-disk-import-pipelines.png "SSP Disk Import Pipelines")
+
+## Common templates: DataSources as DataVolume Source
+
+The common templates shipped with SSP need to have their DataVolumeTemplate sections transition to using a DataSource rather than the inlined PVC source field. This allows the DataImportCron to seamlessly update the underlying PVC the corresponding DataSource points to, which then ensures that the common templates are always using the most up-to-date disk images.
+
+## HCO: Integration with SSP Operator
+
+In order to leverage golden disk image auto delivery/updates at the HCO level, the HCO is responsible for configuring DataImportCrons on the SSP operator’s CR.
+
+For instance, if the HCO wants to develop a feature for automatically sourcing in fedora golden images into a PVC to be used by the SSP's common templates, then the HCO needs to configure the location where the fedora image is sourced from on the SSP CR. 
+
+This document doesn't include within its scope how the HCO should expose enabling the image auto delivery/update mechanism on the HCO API or where the containerDisk images are sourced.
+
+## OCP UI: Integration with DataSources
+
+The UI currently detects if a template has a boot source by checking if the DataVolumeTemplate’s source PVC is present. The UI needs to now be capable of checking if a DataVolumeTemplate’s DataSource has the a ready condition equal to 'true'.
```

Comment:
Sorry for the late response, currently the only info we can show is PVC exist or not, but with this new mechanism, new data may interest the UI user.

Is "Ready" / "Not ready" the only information we want to propagate to the UI users ?
Do we want to expose dates of last poll ?
for exmplae:
"Ready" uploaded 365 days ago
Do we have other states ?
for example:
"Ready" and a new image is failing to upload?
"Ready" and a new image is currently uploading ?
Do we want to expose this states to the UI user ?

Reply:
> Is "Ready" / "Not ready" the only information we want to propagate to the UI users ?
Do we want to expose dates of last poll ?
for exmplae:
"Ready" uploaded 365 days ago
Do we have other states ?
for example:
"Ready" and a new image is failing to upload?
"Ready" and a new image is currently uploading ?
Do we want to expose this states to the UI user ?


I think initially the goal should be feature parity with how the PVC detection works. 

There is indeed some interesting new info that can be provided though. How would you feel about re-approaching some of this as a next iteration after the initial feature lands?

Reply:
> as a next iteration

hmm, I think the requirement for the UI to show info about the `DataImportCron` status, will come up very soon.
I may be missing something, from what I understand, I can find the `DataSource` connected to my templates immediately, but when I want to fetch data from the  `DataImportCron` connected this `DataSource` I don't have a clear path.
In order for the UI to fetch data from the  `DataImportCron` that manages a `DataSource` we need some linkage field, `MnagedBy` ? , `Owner` ? to help the UI find the `DataImportCron` that owns this `DataSource` and fetch some nicer info about the image, WDYT ?

Reply:
Yep, we can use an owner reference for the linkage.

Reply:
Sounds good to me, we should specify in the doc that `DataSource` should be owned by `DataImportCron` for the UI to be able to find them, it will also be a nice way to prevent two `DataImportCron` managing the same `DataSource`, or having orphan `DataSource` with no `DataImportCron`

Reply:
Just to clarify:
At some point we defined that SSP will create DataSources that do not have a Cron managing them yet (if I remember correctly).
If a user toggles the autoUpdate feature we'll create or delete the crons respectively. Using ownerRef will cause the GC to cleanup the DataSources then (on deletion) which might not be intended. SSP can use a different deletionPropagationPolicy, but I'm not sure if this is enough.
What about the Cron setting it's reference into status when updating the import status/target pvc?

Also from the UI perspective I don't know if we need the full one-way relationship or if that's something the UI can build itself as it would probably have both resources loaded at some point. I probably should get more familiar with how the UI works at some point before assuming like this...

I don't remember if we discussed that validation should make sure there is only one Cron per Source, if yes we could also enforce the same name for both resources which would make some validation and also linking easier as we're forcing the same namespace anyways.

Reply:
> Also from the UI perspective I don't know if we need the full one-way relationship:

The UI show a list of `Templates` and want to show for each `Template` data currently only available in the `DataImportCron` for example `ImportSuccessful` condition.

Templates List

Template Name | ImportSuccessful.status | ImportSuccessful.message
---------|-----------------|-------------
template-fedora | "False" | "Unable to reach docker://quay.io/kubevirt/fedora"

If we don't have a link between a `DataSource` and the `DataImportCron` that managed it, the search for the `DataImportCron` that has an `managedDataSource` with the same name, even optimized search (hashed), will slow down rendering of large tables.

btw/ ownership is not the only solution here a `menegedBy` status can also help here.

Reply:
> If a user toggles the autoUpdate feature we'll create or delete the crons respectively. 

Objects can have multiple owner references, but only one reference can have `controller: true` [1].  For example...

```
  ownerReferences:
   - apiVersion: kubevirt.io/madeup
     controller: true
     blockOwnerDeletion: false
     kind: MadeUp
     name: resourceA
     uid: e0c84798-986e-437e-835b-840da938e3f7
   - apiVersion: kubevirt.io/madeup
     controller: false
     blockOwnerDeletion: false
     kind: MadeUp
     name: resourceB
     uid: dc4caedd-bd0e-4af0-99ef-a146cfb19699
```

Both resourceA and resourceB have ownership of the object. It doesn't matter which one (resourceA or resourceB) gets deleted first, the owned object isn't going to get garbage collected until both owners are gone. 

So, both SSP CR and DataImportCron can set ownership references on a DataSource, but only the original creator of the DataSource should set the ownership reference with `controller: true`

Does this make sense? If so, I can capture this in the doc so we're sure we'll implement who can take the `controller:true` ref correctly.


1. https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/controller-ref.md#the-three-laws-of-controllers

Reply:
I updated the doc with what we discussed here regarding multiple owner references.

Reply:
lgtm from the UI side